src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -281,7 +304,7 @@ async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> p
 
         private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
         {
-            Stopped();
+            Dispose();
         }
```

Comment:
not your fault.  but i find it super hard to understnd the lifetimes here and the impact of this stuff.  if the Rpc is diconnected, we dispose ourselves (not our caller).  But what impact does that have on the caller?  What sort of races/invariants are there with these types and how they are used/mutate under the covers.  Would be nice to make cleaner and simpler to understand in the future.

Reply:
Definitely. I don't know all the details either. This will get cleaned up once we move to `IServiceBroker`. https://github.com/dotnet/roslyn/issues/40512

Reply:
#40512 sounds like it will be a *huge* help!